File: KthSmallestElementInABST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        
        ArrayList<Integer> result = new ArrayList<>();
        
        inorder(root, result);
        
        return result.get(k-1);
        
    }
    
    public void inorder(TreeNode root, ArrayList<Integer> result){
        
        if(root == null){
            return;
        }
        
        inorder(root.left, result);
        result.add(root.val);
        inorder(root.right, result);
        
    }
    
}

File: KthSmallestElementInABST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
// TC: O(N) SC: O(N)
var kthSmallest = function(root, k) {
    let count = 0;
    const stack = [];
    let current = root;
    while(current || stack.length) {
        while(current) {
            stack.push(current);
            current = current.left;
        }
        current = stack.pop();
        count++;
        if(count === k) return current.val;
        current = current.right;
    }
    return -1;
};

// TC: O(N) SC: O(N)
var kthSmallest = function(root, k) {
    let result = 0;
    let counter = 0;
    const modifiedInorder = (root) => {
        if(!root) return;
        modifiedInorder(root.left);
        counter++;
        if(counter === k) {
            result = root.val;
            return;
        }
        modifiedInorder(root.right);
    }
    modifiedInorder(root);
    return result;
};
